File: Kern/Sources/DuoKern/FSRS/FSRS.swift
import Foundation

/// FSRS-5 scheduler math. Pure functions over `MemoryState`; no clocks or dates —
/// callers pass elapsed time in (possibly fractional) days.
///
/// Formulas follow the reference implementation ts-fsrs v4.7.1
/// (`src/fsrs/algorithm.ts`, "FSRS-5.0") with short-term (same-day) reviews enabled.
public struct FSRS: Sendable {
    public let parameters: FSRSParameters

    /// Power-law forgetting-curve decay. Fixed in FSRS-5 (trainable only in FSRS-6).
    static let decay: Double = -0.5
    /// factor = 0.9^(1/decay) − 1, so R(t = S) = 0.9 exactly.
    static let factor: Double = 19.0 / 81.0
    static let minStability: Double = 0.01
    static let maxStability: Double = 36500.0
    /// Initial stability lower bound (reference clamps S0 to ≥ 0.1).
    static let minInitialStability: Double = 0.1

    public init(parameters: FSRSParameters = .default) {
        self.parameters = parameters
    }

    // MARK: - Public API

    /// Memory state after the very first review of a card.
    public func initialState(rating: Rating) -> MemoryState {
        MemoryState(
            stability: initialStability(rating),
            difficulty: initialDifficulty(rating)
        )
    }

    /// Memory state after a subsequent review.
    ///
    /// `elapsedDays < 1` takes the short-term (same-day) path:
    /// stability changes by `e^{w17·(G−3+w18)}` and the forgetting curve is not applied,
    /// matching the reference's same-day review handling (t == 0 in whole-day units).
    public func nextState(state: MemoryState, elapsedDays: Double, rating: Rating) -> MemoryState {
        let s = max(state.stability, Self.minStability)
        let d = clampDifficulty(state.difficulty)
        let t = max(elapsedDays, 0)

        let newStability: Double
        if t < 1 {
            newStability = shortTermStability(stability: s, rating: rating)
        } else if rating == .again {
            let r = forgettingCurve(elapsedDays: t, stability: s)
            // why: cap post-lapse stability so a lapse can never leave the card
            // stronger than the same-day-again outcome would (reference behavior).
            let sameDayFloor = s / exp(parameters.w[17] * parameters.w[18])
            newStability = min(
                max(sameDayFloor, Self.minStability),
                forgetStability(difficulty: d, stability: s, retrievability: r)
            )
        } else {
            let r = forgettingCurve(elapsedDays: t, stability: s)
            newStability = recallStability(difficulty: d, stability: s, retrievability: r, rating: rating)
        }
        return MemoryState(
            stability: newStability,
            difficulty: nextDifficulty(d, rating: rating)
        )
    }

    /// Probability of recall after `elapsedDays` given the card's stability. In (0, 1].
    public func retrievability(state: MemoryState, elapsedDays: Double) -> Double {
        forgettingCurve(
            elapsedDays: max(elapsedDays, 0),
            stability: max(state.stability, Self.minStability)
        )
    }

    /// Next review interval in whole days (reference rounds half-up),
    /// clamped to `1...maximumInterval`.
    public func nextIntervalDays(stability: Double, desiredRetention: Double? = nil) -> Double {
        let retention = desiredRetention ?? parameters.desiredRetention
        precondition(retention > 0 && retention <= 1, "desiredRetention must be in (0, 1]")
        // I(r, s) = s · (r^{1/decay} − 1) / factor; equals s when r = 0.9.
        let modifier = (pow(retention, 1 / Self.decay) - 1) / Self.factor
        let interval = (stability * modifier).rounded(.toNearestOrAwayFromZero)
        return min(max(interval, 1), parameters.maximumInterval)
    }

    /// Card phase transition on review (reference basic-scheduler state machine).
    public func nextPhase(current: CardPhase, rating: Rating) -> CardPhase {
        switch current {
        case .new:
            return rating == .easy ? .review : .learning
        case .learning, .relearning:
            return (rating == .good || rating == .easy) ? .review : current
        case .review:
            return rating == .again ? .relearning : .review
        }
    }

    // MARK: - FSRS-5 core formulas (internal for golden-vector tests)

    /// R(t, S) = (1 + factor · t / S)^decay
    func forgettingCurve(elapsedDays: Double, stability: Double) -> Double {
        pow(1 + Self.factor * elapsedDays / stability, Self.decay)
    }

    /// S0(G) = w_{G−1}, clamped to ≥ 0.1.
    func initialStability(_ rating: Rating) -> Double {
        max(parameters.w[rating.rawValue - 1], Self.minInitialStability)
    }

    /// D0(G) = w4 − e^{(G−1)·w5} + 1, unclamped (mean reversion uses the raw value).
    func rawInitialDifficulty(_ rating: Rating) -> Double {
        parameters.w[4] - exp(Double(rating.rawValue - 1) * parameters.w[5]) + 1
    }

    func initialDifficulty(_ rating: Rating) -> Double {
        clampDifficulty(rawInitialDifficulty(rating))
    }

    /// Linear damping ΔD·(10−D)/9, then mean reversion toward D0(easy):
    /// D′ = w7·D0(4) + (1−w7)·(D + ΔD·(10−D)/9) with ΔD = −w6·(G−3).
    func nextDifficulty(_ difficulty: Double, rating: Rating) -> Double {
        let deltaD = -parameters.w[6] * Double(rating.rawValue - 3)
        let damped = difficulty + deltaD * (10 - difficulty) / 9
        let reverted = parameters.w[7] * rawInitialDifficulty(.easy)
            + (1 - parameters.w[7]) * damped
        return clampDifficulty(reverted)
    }

    /// S′_r = S·(1 + e^{w8}·(11−D)·S^{−w9}·(e^{w10·(1−R)}−1)·hardPenalty·easyBonus)
    func recallStability(
        difficulty: Double, stability: Double, retrievability: Double, rating: Rating
    ) -> Double {
        let hardPenalty = rating == .hard ? parameters.w[15] : 1
        let easyBonus = rating == .easy ? parameters.w[16] : 1
        let growth = exp(parameters.w[8])
            * (11 - difficulty)
            * pow(stability, -parameters.w[9])
            * (exp((1 - retrievability) * parameters.w[10]) - 1)
            * hardPenalty
            * easyBonus
        return clampStability(stability * (1 + growth))
    }

    /// S′_f = w11·D^{−w12}·((S+1)^{w13}−1)·e^{w14·(1−R)}
    func forgetStability(difficulty: Double, stability: Double, retrievability: Double) -> Double {
        clampStability(
            parameters.w[11]
                * pow(difficulty, -parameters.w[12])
                * (pow(stability + 1, parameters.w[13]) - 1)
                * exp((1 - retrievability) * parameters.w[14])
        )
    }

    /// Same-day review: S′_s = S·e^{w17·(G−3+w18)}
    func shortTermStability(stability: Double, rating: Rating) -> Double {
        clampStability(
            stability * exp(parameters.w[17] * (Double(rating.rawValue) - 3 + parameters.w[18]))
        )
    }

    private func clampDifficulty(_ difficulty: Double) -> Double {
        min(max(difficulty, 1), 10)
    }

    private func clampStability(_ stability: Double) -> Double {
        min(max(stability, Self.minStability), Self.maxStability)
    }
}

File: kern/src/commonMain/kotlin/net/spross/kern/fsrs/Fsrs.kt
package net.spross.kern.fsrs

import kotlin.math.exp
import kotlin.math.floor
import kotlin.math.max
import kotlin.math.min
import kotlin.math.pow
import net.spross.kern.model.MemoryState
import net.spross.kern.model.Rating

/**
 * FSRS-6 formula set. Pure math over [MemoryState]; no clocks — callers pass
 * elapsed time in fractional days (`< 1.0` takes the same-day short-term path).
 *
 * Formulas follow ts-fsrs v5.4.1 / py-fsrs v6.3.1 with the divergences resolved
 * toward ts-fsrs/fsrs-rs: same-day `sinc >= 1` mask for Hard and above, and the
 * S_MIN 0.001 initial-stability floor (ts-only 0.1 floor not adopted).
 */
class Fsrs(val parameters: FsrsParameters = FsrsParameters()) {

    private val w: List<Double> = parameters.w

    /** decay = −w20; factor makes R(t = S) = 0.9 exactly. */
    private val decay: Double = -w[20]
    private val factor: Double = 0.9.pow(1.0 / decay) - 1.0

    /** R(t, S) = (1 + factor · t / S)^decay — probability of recall, in (0, 1]. */
    fun retrievability(elapsedDays: Double, stability: Double): Double {
        val s = max(stability, S_MIN)
        val t = max(elapsedDays, 0.0)
        return (1.0 + factor * t / s).pow(decay)
    }

    /**
     * Memory state after a review. `memory == null` means first review
     * (initial S0/D0); `elapsedDays < 1.0` takes the short-term path.
     */
    fun nextMemory(memory: MemoryState?, elapsedDays: Double, rating: Rating): MemoryState {
        if (memory == null) {
            return MemoryState(
                stability = initialStability(rating),
                difficulty = initialDifficulty(rating),
            )
        }
        val s = max(memory.stability, S_MIN)
        val d = clampDifficulty(memory.difficulty)
        val t = max(elapsedDays, 0.0)
        val stability = when {
            t < 1.0 -> shortTermStability(s, rating)
            rating == Rating.Again -> {
                // why: cap post-lapse stability at the same-day-again outcome so a lapse
                // can never leave the card stronger than failing it immediately would.
                val ceiling = s / exp(w[17] * w[18])
                min(max(ceiling, S_MIN), forgetStability(d, s, retrievability(t, s)))
            }
            else -> recallStability(d, s, retrievability(t, s), rating)
        }
        return MemoryState(stability = stability, difficulty = nextDifficulty(d, rating))
    }

    /**
     * I(rd, S) = S / factor · (rd^(1/decay) − 1), in fractional days and
     * unclamped — the interval the model actually asks for. Equals S at
     * rd = 0.9. [FsrsScheduler] quantizes and clamps it per parameters.
     */
    fun intervalRawDays(
        stability: Double,
        desiredRetention: Double = parameters.desiredRetention,
    ): Double {
        require(desiredRetention > 0.0 && desiredRetention <= 1.0) {
            "desiredRetention must be in (0, 1]"
        }
        return stability / factor * (desiredRetention.pow(1.0 / decay) - 1.0)
    }

    /**
     * [intervalRawDays] rounded half-up to whole days, then clamped to
     * `[1, maximumIntervalDays]` — the reference day-bucket convention.
     */
    fun intervalDays(
        stability: Double,
        desiredRetention: Double = parameters.desiredRetention,
    ): Int {
        val rounded = floor(intervalRawDays(stability, desiredRetention) + 0.5)
        return min(max(rounded, 1.0), parameters.maximumIntervalDays.toDouble()).toInt()
    }

    // Formula internals (exposed to the module's tests for the reference unit vectors).

    /** S0(G) = w[G−1], floored at S_MIN (py-fsrs/fsrs-rs convention). */
    internal fun initialStability(rating: Rating): Double =
        max(w[rating.value - 1], S_MIN)

    /** D0(G) = w4 − e^{w5·(G−1)} + 1, unclamped (mean reversion anchors on the raw value). */
    internal fun rawInitialDifficulty(rating: Rating): Double =
        w[4] - exp(w[5] * (rating.value - 1)) + 1.0

    internal fun initialDifficulty(rating: Rating): Double =
        clampDifficulty(rawInitialDifficulty(rating))

    /** ΔD = −w6·(G−3) with linear damping, then mean reversion toward UNCLAMPED D0(Easy). */
    internal fun nextDifficulty(difficulty: Double, rating: Rating): Double {
        val deltaD = -w[6] * (rating.value - 3)
        val damped = difficulty + deltaD * (10.0 - difficulty) / 9.0
        val reverted = w[7] * rawInitialDifficulty(Rating.Easy) + (1.0 - w[7]) * damped
        return clampDifficulty(reverted)
    }

    /** S′_r = S·(1 + e^{w8}·(11−D)·S^{−w9}·(e^{w10·(1−R)}−1)·hardPenalty·easyBonus) */
    internal fun recallStability(
        difficulty: Double,
        stability: Double,
        retrievability: Double,
        rating: Rating,
    ): Double {
        val hardPenalty = if (rating == Rating.Hard) w[15] else 1.0
        val easyBonus = if (rating == Rating.Easy) w[16] else 1.0
        val growth = exp(w[8]) *
            (11.0 - difficulty) *
            stability.pow(-w[9]) *
            (exp(w[10] * (1.0 - retrievability)) - 1.0) *
            hardPenalty *
            easyBonus
        return clampStability(stability * (1.0 + growth))
    }

    /** S′_f = w11·D^{−w12}·((S+1)^{w13}−1)·e^{w14·(1−R)} */
    internal fun forgetStability(
        difficulty: Double,
        stability: Double,
        retrievability: Double,
    ): Double = clampStability(
        w[11] *
            difficulty.pow(-w[12]) *
            ((stability + 1.0).pow(w[13]) - 1.0) *
            exp(w[14] * (1.0 - retrievability)),
    )

    /**
     * Same-day: S′_s = S · sinc with sinc = S^{−w19}·e^{w17·(G−3+w18)}.
     * why: sinc is clamped >= 1 for G >= Hard (fsrs-rs/ts-fsrs semantics; py-fsrs
     * v6.3.1 masks only Good/Easy — resolved toward the canonical algorithm source).
     */
    internal fun shortTermStability(stability: Double, rating: Rating): Double {
        val sinc = stability.pow(-w[19]) * exp(w[17] * (rating.value - 3 + w[18]))
        val masked = if (rating.value >= Rating.Hard.value) max(sinc, 1.0) else sinc
        return clampStability(stability * masked)
    }

    private fun clampDifficulty(difficulty: Double): Double = min(max(difficulty, 1.0), 10.0)

    private fun clampStability(stability: Double): Double = min(max(stability, S_MIN), S_MAX)

    companion object {
        const val S_MIN: Double = 0.001
        const val S_MAX: Double = 36500.0
    }
}
